### ISSUE TITLE
Evaluate Kuery async changes on TechEmpower at scale

### ISSUE THREAD

**Issue by djones6** [IBM-Swift/Kitura-Benchmarks#11]
### Context & Description
Evaluate the asynchronous changes to Kuery and Kuery-PostgreSQL using the [updated TechEmpower benchmarks](https://github.com/IBM-Swift/Kitura-Benchmarks/pull/7).

This will allow the `/queries` and `/updates` routes to function recursively, rather than the current workaround of using `/updatesParallel` and `/queriesParallel` (executing multiple finds in parallel, and using a semaphore to keep track of when all have completed - which blocks a thread).

### Background
The issue with the recursive definition for updates and queries is that, because closures are currently invoked synchronously, the connection used for query N is not returned to the pool until the end of that closure - unavailable for query N+1 (invoked from within the closure).

For queries > 1 and many concurrent clients, eventually the benchmark will deadlock, because there are a number of clients waiting for a connection that they can never have because the pool has become empty, and preventing the connections that were just used from returning to the pool.

eg:
```
djones6@gruffalo:~$ wrk -c256 -t2 -d10s -i1s http://localhost:8080/queries?queries=3
Running 10s test @ http://localhost:8080/queries?queries=3
  2 threads and 256 connections
   1s: 7337 requests in last 1s (   7337.0 req/sec)
   2s: 7501 requests in last 1s (   7501.0 req/sec)
   3s: 7524 requests in last 1s (   7524.0 req/sec)
   4s: 7472 requests in last 1s (   7472.0 req/sec)
   5s: 7434 requests in last 1s (   7434.0 req/sec)
   6s: 7369 requests in last 1s (   7369.0 req/sec)
   7s: 7203 requests in last 1s (   7203.0 req/sec)
   8s: 6991 requests in last 1s (   6991.0 req/sec)
   9s: 80 requests in last 1s (     80.0 req/sec)
  10s: 0 requests in last 1s (      0.0 req/sec)
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    33.53ms    3.87ms 234.91ms   81.76%
    Req/Sec     3.68k   296.36     4.03k    95.65%
  58911 requests in 10.03s, 256 connections, 15.08MB read
Requests/sec:   5873.08
Requests/conn:   230.12
Transfer/sec:      1.50MB
```

## Clear & Actionable Tasks
- [ ] Test TechEmpower using https://github.com/IBM-Swift/Swift-Kuery-PostgreSQL/tree/issue_async
- [ ] Evaluate queries and updates under stress, up to 20 queries and 256 concurrent clients

### Exit Criteria:
This issue is complete once the following criteria are met:
+ Code...
+ Tests...

**Comment by djones6** [comment]
The current `issue_async` changes for Kuery and KueryPostgres do not behave well under load when there are more clients than available Dispatch threads (64 on Mac). 

@kilnerm and I analyzed the behaviour.  In this scenario, all threads eventually become blocked waiting for a connection from the connection pool, because the (offloaded) execute calls do not have a thread available to run on, but are holding connections.

The solution we think is to make `ConnectionPool.getConnection()` asynchronous, where its completion will execute immediately if a connection is available, otherwise it will be enqueued and executed when a connection returns to the pool.

**Comment by djones6** [comment]
Yesterday we tested the latest iteration of the `issue_async` changes, including the redesigned `ConnectionPool.getConnection(completion:)` with task queueing.

These changes worked very well on macOS under Xcode, with the queue operating as expected, without any serious performance issue (in debug mode under Xcode).

Next step is to repeat the stress test on Linux in release mode, and get some sense of how it performs.

**Comment by djones6** [comment]
Changes passed testing on Linux and have been merged into their respective `next` branches.

The changes for the ORM are ready to test, and I will be doing so today.

**Comment by djones6** [comment]
ORM changes (including async result fetcher changes) all work correctly under stress on Mac and Linux.  Benchmark changes are in this branch: https://github.com/IBM-Swift/Kitura-Benchmarks/tree/issue_async_results

Once Kuery 3.0 is tagged, these changes can be pushed up to the TechEmpower project.
